src/main/java/im/status/keycard/KeycardApplet.java

```diff
@@ -1076,12 +1138,31 @@ private void generateKey(APDU apdu) {
       ISOException.throwIt(ISO7816.SW_CONDITIONS_NOT_SATISFIED);
     }
 
-    apduBuffer[ISO7816.OFFSET_LC] = BIP39_SEED_SIZE;
+    // Sanitize P1 input
+    byte exportability = (byte) MASTERSEED_EMPTY;
+    switch(apduBuffer[ISO7816.OFFSET_P1]) {
+      case (byte) 0:
+        exportability = (byte) MASTERSEED_NOT_EXPORTABLE;
+        break;
+      case (byte) 1:
+        exportability = (byte) MASTERSEED_EXPORTABLE;
+        break;
+      default:
+        ISOException.throwIt(ISO7816.SW_INCORRECT_P1P2);
+        break;
+    }
+
+    // Generate random seed
     crypto.random.generateData(apduBuffer, ISO7816.OFFSET_CDATA, BIP39_SEED_SIZE);
 
+    // Load the generated seed
     loadSeed(apduBuffer);
     pinlessPathLen = 0;
-    generateKeyUIDAndRespond(apdu, apduBuffer);
+
+    // Save seed exportability (this also indicates the seed as valid)
+    masterSeedStatus = exportability; // Only save seed exportability after seed successfully loaded
+
+    secureChannel.respond(apdu, BIP39_SEED_SIZE, ISO7816.SW_NO_ERROR);
```

Comment:
Ok, the only other thing I would consider is if we should modify the "non-exportable" case so that it doesn't export.

Not sure if you guys want: "never exportable ever"
or just: "export once at creation, but _never again_ exportable"

Reply:
If we ant safecards to be usable with insecure environments (ie user's laptop), we may want a creation mechanism that doesn't even export the seed at creation. IMO this would be most compatible with existing keycard tech (Ledger, Yubikey, etc)

Reply:
I like the idea of expanding the flag 👍

Reply:
1445c98 (will squash later for feature cleanup)